### ISSUE TITLE
last_sync_date

### ISSUE THREAD

**Issue by luigi37** [abeauseigle/TodoApp#13]
Hi Alain,
I'm facing a quite nasty issue which I can't understand with last_sync_date 

My server migrated to USA so timezone in -4:00

My app is in Italy so +2:00

Now, if I modify both in the App and on the server, the App always win because... last_sync_date is a total of +6 hours ahead...

Am I missing anything? I mean... I assumed the sync function was timezone proof but now I'm wondering if I was wrong....

Luigi

**Comment by luigi37** [comment]
Ok it looks it's an issue in my configuration. I migrated my server from Italy to USA and there is a discrepancy due to my settings: mysql is on USA timezone, PHP was set up by me on IT...
That means that last_sync_date put automatically by mysql was behind 6 hours, so it was looking older than the change on the app...

I discovered then that the websql takes the lastsyncdate from the server... so once I corrected my last dates (after putting mysql and php on same timezone) all was good...

Only question... the server lastsyncdate is based on the highest date for the set of info or on the single record date?

Thanks!

**Comment by sachin0606** [comment]
Hi. 

i am trying to do this but i am unable to get last sync date , first time data is not getting downloaded, can you tell me what might be the issue.

**Comment by luigi37** [comment]
Did you try forcing lastsyncdate client side to zero? That force a first time resync

**Comment by sachin0606** [comment]
yes i did, but first time the data is not getting downloaded from server.

**Comment by luigi37** [comment]
Do you mean in subsequent changes it works?

**Comment by abeauseigle** [comment]
Hi Sachin,

Did you test your php code without client side code? 
Go in the getModified___.php code and you will see come comments about testing it (Around line 7 to 9).
Go in the webSqlSyncAdapter.php (line 21 to 25). A $BDBid == 0; will force a first sync.
3 years ago, it helped me a lot to debug my server side code.
Check the server answer (json).

Alain from Montréal

**Comment by abeauseigle** [comment]
Hi Luigi, 
About your question... the server lastsyncdate is based on the highest date for the set of info or on the single record date? 

If you see in GetModifiedTodo.php, you will see 
AND Todo.TodoLastModifDateH > '" .$clientLastSyncDate. "'
So it get the last modified for each record.

I use the $clientLastSyncDate (dateTime of the last sync saved client side but coming from the server) to identify CONFLICTs (modified both side between 2 sync). If we have a conflict, Server will win but the client will receive a message about it. 

ALL DateTimes come from the server, NEVER from the client.

Note: It's very far in my memory (almost 3 years ago, I did it).

A+

**Comment by abeauseigle** [comment]
Hi Sachin,

After you gonna do the first sync, I suggest using the "debug1" and "debug2" variables to debug server side if you have any server problems. See webSqlSyncAdapter.php line 43. If not, we are blind for debugging server side (queries particularly). It's usefull because, we cannot use a Php Echo like in a standard php application. 
ex: In TodoSet.php
			//$debug1 = "Line 24 Count: ".$count; // Show the query in the Return answer for debugging purpose
			//$debug2 = "Line 25"; // Show the query in the Return answer for debugging purpose

You can see the server answer in the browser console.

Good lock.

**Comment by luigi37** [comment]
> ALL DateTimes come from the server, NEVER from the client.

Thanks Alain.

I have a server VS server issue... 

Now I have MYSQL server on EST timezone and PHP on CET so I have 6 hours difference....

I realised all my records are based on the fact that the server data is assumed to be Europe/Rome so changing PHP timezone is going to be a pain which I would like to avoid.

What's happening now is that the lastsyncdate driven by the modify trigger is 6 hours behind compared to the time returned by PHP.

Outcome is that the client always win because it gets back from PHP server Rome time, while the check in MYSQL database last_sync_date is in US time...

I double checked by manually changing by 6 hours last_sync_date and it works as expected...

To overcome this I tried to change in wss_adapter: 

    $clientLastSyncDate = $handler -> clientData['info']['lastSyncDate']/1000

to

    $clientLastSyncDate = $handler -> clientData['info']['lastSyncDate']/1000 - 6*3600

but it seems it's not enough....

Any suggestion?

**Comment by luigi37** [comment]
I think I've got it..

Adapter:

   function myJob($handler){
	$currentDateTime =  date("Y-m-d H:i:s");
	$clientLastSyncDate = $handler -> clientData['info']['lastSyncDate']/1000 - 6*3600;

This last $clientLastSyncDate in practive is never used AND it's not coherent with timestamp (because it's unixtime)

So... where is set $clientLastSyncDate?

It's set in "set....php":

I changed both

    	$currentDateTime =  date("Y-m-d H:i:s", strtotime('-6 hour'));
	$clientData = $handler -> clientData;

	// $BDBid = $clientData['info']['BDBid'];	// Browser DB unique id (attributed at the First sync). Equal to the first sync DateTime in Unix format.
	$clientLastSyncDateUnix= $clientData['info']['lastSyncDate']/1000 - 6*3600;
        $clientLastSyncDate= date('Y-m-d H:i:s', $clientLastSyncDateUnix);

So... the $clientLastSyncDate is used both in set and in get AND $currentDateTime is used when setting up the updated record as "last_sync_date"

It seems it works...

Any thought?

PS: Why did you set up specifically last_sync_date when it could have been setup by the "modify trigger"?

**Comment by luigi37** [comment]
Finally, to make it future proof for next server change I get the delta automatically from the mysql and php servers:

        $sql = "SELECT NOW() as curtime";
        $result = $db->sql_query($sql) or die(mysql_error($db));
        $row = $db->sql_fetchrow($result);
        $deltaTimeMysqlPHP=round((strtotime($row['curtime'])-strtotime(date("Y-m-d H:i:s")))/3600);
        

	$currentDateTime =  date("Y-m-d H:i:s", strtotime($deltaTimeMysqlPHP.' hour'));
	$clientData = $handler -> clientData;

	// $BDBid = $clientData['info']['BDBid'];	// Browser DB unique id (attributed at the First sync). Equal to the first sync DateTime in Unix format.
	$clientLastSyncDateUnix= $clientData['info']['lastSyncDate']/1000 + $deltaTimeMysqlPHP *3600;	// It gives a 10 digits Unix dateTime format
	$clientLastSyncDate= date('Y-m-d H:i:s', $clientLastSyncDateUnix);	// to show the date in YYYY-MM-DD HH:MM:SS format (MySQL datetime format). Result: 2007-12-20 14:00:00

**Comment by abeauseigle** [comment]
Great Luigi,
It's very interesting. And more is your $deltaTimeMysqlPHP to solve the problem automatically.
I think about the $currentDateTime  =  date("Y-m-d H:i:s"); in setTodo.php could be placed in comment because it's written after  // usefull for the unit test of this function.  

For your question: Why did you set up specifically last_sync_date when it could have been setup by the "modify trigger"? 
Really, I don't remember. May be I took this part from Orbitaloop.

See you.

**Comment by sachin0606** [comment]
Hi 
@abeauseigle / @luigi37:

below is the error message i am getting.


line 7username : myusername password : test

line 10 username is myusername password is test

Line 34 ResourceID: userUserID: CLnum_rows: 0 BDBID: Your're not authorized to get the data from this server.
Warning: session_destroy(): Trying to destroy uninitialized session in F:\xampp\htdocs\TodoApp2\webSqlSyncAdapter.php on line 56 


as per my knowladge i am unable to get BDBid.


any issue??
